### ISSUE TITLE
asterisk VIRTUAL_HOST doesn't work with latest nginx-proxy

### ISSUE THREAD

**Issue by sandrodz** [jwilder/nginx-proxy#720]
I'm using `*.domain.com` as `VIRTUAL_HOST` but it fails on recent version of nginx-proxy, it worked before just fine, so I had to roll back to previous version.

Version that works fine is from 5 month ago, idk exact version: 
<img width="683" alt="screen shot 2017-02-15 at 11 12 01 am" src="https://cloud.githubusercontent.com/assets/8479569/22964036/e8f96df0-f36f-11e6-8afc-ae04708799d1.png">

Any ideas what happened in new version?

**Comment by thomasleveil** [comment]
Please provide more info as explained in https://github.com/jwilder/nginx-proxy/wiki/Troubleshooting

**Comment by hiroponz** [comment]
I have faced the same problem.
The problem was same with #580.

**Comment by sandrodz** [comment]
@hiroponz I don't see how these are related. Besides I've problem with the latest version, where fix seems to have been merged. Care to explain in detail your problem and solution?

**Comment by hiroponz** [comment]
In my case I have mounted `/etc/nginx/conf.d` and placed `redirect.conf` to redirect, for example, `https://example.com` to `https://www.example.com`.
I had to add `ssl_session_tickets off;` to the `redirect.conf`.

See also https://community.letsencrypt.org/t/errors-from-browsers-with-ssl-session-tickets-off-nginx/18124/5.

**Comment by sandrodz** [comment]
@hiroponz but it was merged as off in latest version. I've problem with latest version.

@thomasleveil 

diagnostic info as requested, some logs etc.

with *.domain.com and latest nginx-proxy:

<img width="1624" alt="screen shot 2017-02-16 at 2 28 03 pm" src="https://cloud.githubusercontent.com/assets/8479569/23018084/27659866-f456-11e6-96a2-f0e5beb83bc7.png">

in the browser:
<img width="1323" alt="screen shot 2017-02-16 at 2 28 29 pm" src="https://cloud.githubusercontent.com/assets/8479569/23018120/426893fc-f456-11e6-96dd-3158dcccf530.png">

curl:
<img width="518" alt="screen shot 2017-02-16 at 2 28 53 pm" src="https://cloud.githubusercontent.com/assets/8479569/23018141/553f626c-f456-11e6-9a45-d5ddced44945.png">

letsencrypt companion:
<img width="861" alt="screen shot 2017-02-16 at 2 29 09 pm" src="https://cloud.githubusercontent.com/assets/8479569/23018164/68362220-f456-11e6-9d92-955f98370fed.png">

I think this last log from companion container provides insight into the nature of the problem. If I use older nginx-proxy all works fine.

**Comment by thomasleveil** [comment]
What about the result of `docker exec nginxproxy nginx - s reload`?

**Comment by sandrodz** [comment]
@thomasleveil 

<img width="763" alt="screen shot 2017-02-16 at 5 04 32 pm" src="https://cloud.githubusercontent.com/assets/8479569/23022423/208a7ba6-f46a-11e6-9d35-f5714fcb7e27.png">

**Comment by sandrodz** [comment]
@thomasleveil 

this is **working** config in old nginx-proxy:

```
upstream
*.weare.de.com
 {
                                ## Can be connect with "bridge" network
                        # webserver
                        server 172.17.0.8:80;
}
server {
        server_name
*.weare.de.com
;
        listen 80 ;
        access_log /var/log/nginx/access.log vhost;
        include /etc/nginx/vhost.d/default;
        location / {
                proxy_pass http://*.weare.de.com;
        }
}

```

and this is **not working** config in latest nginx-proxy

**Comment by sandrodz** [comment]
its obvious config file is not generated correctly in the recent version when using asterisk in the domain.

**Comment by thomasleveil** [comment]
changes were merged yesterday. Give it another try

**Comment by sandrodz** [comment]
@thomasleveil I pulled the latest and problem remains unfortunately.

**Comment by sandrodz** [comment]
@thomasleveil hey, any thoughts? latest version fails for me exactly like before...

**Comment by sandrodz** [comment]
@jwilder could you please chime in? I'm stuck on old version and cannot update.

**Comment by jwilder** [comment]
Looks like #679 doesn't resolve this since a regex is not being used. We may need to revert the sha upstream change.

**Comment by sandrodz** [comment]
@jwilder how can I help? issue is quite simple to reproduce so...

**Comment by thomasleveil** [comment]
@sandrodz, I struggle to reproduce your issue. By using `VIRTUAL_HOST: "*.domain.com"` in a web container, the latest version of nginx-proxy produces this nginx conf:

    # *.domain.com
    upstream *.domain.com {
    				## Can be connect with "issues_default" network
    			# issues_web_1
    			server 172.22.0.2:81;
    }
    server {
    	server_name *.domain.com;
    	listen 80 ;
    	access_log /var/log/nginx/access.log vhost;
    	location / {
    		proxy_pass http://*.domain.com;
    	}
    }

and my tests with `1.domain.com` or `2.domain.com` are OK.

The only difference in my test and you situation seems to be the `include /etc/nginx/vhost.d/default;` directive. Can you share the full content of the `/etc/nginx/vhost.d/default` file you have?

**Comment by sandrodz** [comment]
@thomasleveil 

I have multiple domains in VIRTUAL_HOST besides the asterisk one.

in default I've:
```
## Start of configuration add by letsencrypt container
location /.well-known/acme-challenge/ {
    allow all;
    root /usr/share/nginx/html;
    try_files $uri =404;
    break;
}
## End of configuration add by letsencrypt container
client_max_body_size 100m;
```

**Comment by thomasleveil** [comment]
> I have multiple domains in VIRTUAL_HOST besides the asterisk one.

Please copy/paste the exact content of this environment variable so we can reproduce. The ideal would be if you could provide a simple docker-compose.yml file that would allow any of us to reproduce the issue.

The fact that you are using the letsencrypt container with nginx-proxy is also an important information since multiple users have issues only when using nginx-proxy with this container. See for instance https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion/pull/168

**Comment by sandrodz** [comment]
@thomasleveil sure, I'll try my best.

So nginx-proxy + certs companion docker-compose.yml looks like this:

```
nginx-proxy:
  image: jwilder/nginx-proxy
  container_name: nginx-proxy
  ports:
    - "80:80"
    - "443:443"
  restart: always
  environment:
    - DEFAULT_HOST=${DEFAULT_HOST}
  volumes:
    - /srv/certs:/etc/nginx/certs:ro
    - /var/run/docker.sock:/tmp/docker.sock:ro
    - ./vhost.d:/etc/nginx/vhost.d
    - /usr/share/nginx/html


nginx-certs:
  image: jrcs/letsencrypt-nginx-proxy-companion
  container_name: nginx-certs
  volumes:
    - /srv/certs:/etc/nginx/certs:rw
    - /var/run/docker.sock:/var/run/docker.sock:ro
  volumes_from:
    - nginx-proxy
```

and webserver config:

```
webserver:
  image: nginx:1.11-alpine
  container_name: webserver
  ports:
    - "127.0.0.1::80"
    - "127.0.0.1::443"
  external_links:
    - php5-fpm
    - php7-fpm
  restart: always
  environment:
    VIRTUAL_HOST: |
      wearede.com,
      www.wearede.com,
      weare.de.com,
      *.weare.de.com
    LETSENCRYPT_HOST: |
      wearede.com,
      www.wearede.com,
      weare.de.com
    LETSENCRYPT_EMAIL: support@wearede.com
  volumes:
    - /srv/www:/www
```

**Comment by thomasleveil** [comment]
are those in 2 separated docker-compose.yml files?

**Comment by sandrodz** [comment]
@thomasleveil yes

**Comment by thomasleveil** [comment]
I have to go, but I'll give a look at this later today

**Comment by sandrodz** [comment]
@thomasleveil thank you! I will be around later today in case I need to provide further info.

**Comment by thomasleveil** [comment]
@sandrodz that was a tricky one, but I got it figured out.

## tl;dr

in your docker-compose file, replace 

```
  environment:
    VIRTUAL_HOST: |
      wearede.com,
      www.wearede.com,
      weare.de.com,
      *.weare.de.com
    LETSENCRYPT_HOST: |
      wearede.com,
      www.wearede.com,
      weare.de.com
```
with
```
  environment:
    VIRTUAL_HOST: >
      wearede.com,
      www.wearede.com,
      weare.de.com,
      *.weare.de.com
    LETSENCRYPT_HOST: >
      wearede.com,
      www.wearede.com,
      weare.de.com
```

## Analysis

In the yaml syntax, the [Block scalar indicator `|`](http://yaml.org/spec/1.2/spec.html#id2773653) preserves new line characters. nginx-proxy breaks when there are new line characters in the value of a `VIRTUAL_HOST` environment variable. (and by the way jrcs/letsencrypt-nginx-proxy-companion has this same restriction for the `LETSENCRYPT_HOST` environment variable).

The yaml _Folded scalar indicator_ `>` removes the new lines characters.

@jwilder in the end,  #671 or #679 are not involved in the current state of this issue. One way to be more forgiving for users would be to have nginx-proxy trim the `$host` values from `VIRTUAL_HOST` as @md5 suggested in #271. In my tests, just adding `{{ $host := trim $host }}` allows for any kind of white character (new line include) to be part of the `VIRTUAL_HOST` environment variable without affecting the validity of the nginx configuration file.

**Comment by sandrodz** [comment]
@thomasleveil wow thank you! it works now.
